### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by aflag**
# Performance improvements

Closes: https://github.com/ScottishCovidResponse/SCRCIssueTracking/issues/639

## Rationale

The main performance improvement was caused by migrating from scipy to numpy for drawing from distributions. However, it is also possible to improve the performance by 30% by reducing the number of calls to `numpy.random.hypergeometric`. Gains can also be made by making similar changes to the calls to `binomial` in `getIncomingInfectiousContactsByNode`, however, surprisingly, `getIncomingInfectiousContactsByNode` very little of CPU time is spent there. Furthermore, changing the multinominal in `getInternalProgressionAllNodes` is not as simple, but there could be some performance to be gained there too.

The change here is to pass arrays to `numpy.random.hypergeometric` rather call it multiple times. This improves the speed because the iteration and computation is all done internally in C inside of hypergeometric, rather than using an outer python loop.

## Flame graph

This is the flame graph before the changes in this PR:

numpy

After the changes, `getInternalInfectiousContacts` is not the function which uses the most CPU cycles, but rather, it becomes `getInternalProgressionAllNodes`. This is the flame graph after the changes:

faster

## Correctness evidence

Furthermore, the changes in this PR does not seem to significantly alter the output of `getInternalInfectiousContacts`. The code below evidences that:

```python
def test_doInternalInfectionProcess_between_ages():
    current_state = {
        ("m", "S"): 20.0,
        ("m", "A"): 150.0,
        ("m", "I"): 300,
        ("o", "S"): 15.0,
        ("o", "A"): 200.0,
        ("o", "I"): 100.0,
    }
    age_matrix = loaders.MixingMatrix({"m": {"m": 0.2, "o": 0.5}, "o": {"o": 0.3, "m": 0.5}})
    a = []
    b = []
    for seed in numpy.random.SeedSequence().generate_state(1000000):
        new_infected = np.getInternalInfectiousContactsInNode(current_state, age_matrix, 0.3, ["I", "A"], True, numpy.random.default_rng(seed))
        a.append(new_infected["m"])
        b.append(new_infected["o"])
    import pickle
    with open("/tmp/x", "wb") as fp:
        pickle.dump((a,b), fp)
```

I used the files saved in `/tmp` to plot the following graphs distributions.

**Before the change**

image
image

* List A: avg: 3.064647 std: 1.750867
* LIst B: avg: 4.499228 std: 2.122463

**After the change**

image
image

List A: avg: 3.061896 std: 1.750817
List B: avg: 4.498371 std: 2.122019

## End to end runs

These are the output of time for runs of 20 trials and 400 nodes.

**Run time using scipy**
real	11m19.108s
user	11m18.132s
sys	0m0.613s

**Run time using numpy**
real	5m29.572s
user	5m28.927s
sys	0m0.572s

**Run time after this PR**
real	4m37.383s
user	4m36.343s
sys	0m0.782s

## Numba

I also tried out [numba](http://numba.pydata.org/), without much success. Numba can only compile functions handling only primitives (lists and dicts have limited support and only when using numba's implementation of those), which is not something we commonly have. I did, however, attempt to convert our inner dict, `Dict[Tuple[Age, Compartment], float]` into a numba dict (numba doesn't support nested dicts). That made the code a lot slower. Even though I could compile some of the functions (not all, as some require nested dicts or objects like `numpy.random.Generator`), just the fact that using pure python with numba dict is a lot slower was sufficient to significantly slow down the entire simulation.

Below is a code I wrote that shows how numba can be inefficient in a case like ours (in particular, notice how much slower pure python + numba dicts is):

```python
from numba import njit
from numba.typed import Dict, List
from numba.core import types


import time

def numba_bench():
    @njit
    def f(d, y, strs):
        for x in strs:
            d[(y,x)] = x

    for y in range(10):
        d=Dict.empty(key_type=types.Tuple((types.string, types.string)), value_type=types.string)
        acc = 0
        ls = List()
        for x in range(100000):
            ls.append(str(x))
        t0 = time.time()
        f(d, str(y), ls)
        acc += time.time() - t0
    print(f"Numba\t{acc}")


def mixed_bench():
    def f(d, y, strs):
        for x in strs:
            d[(y,x)] = x

    for y in range(10):
        d=Dict.empty(key_type=types.Tuple((types.string, types.string)), value_type=types.string)
        acc = 0
        ls = List()
        for x in range(100000):
            ls.append(str(x))
        t0 = time.time()
        f(d, str(y), ls)
        acc += time.time() - t0
    print(f"Mixed\t{acc}")


def python_bench():
    def f(d, y, strs):
        for x in strs:
            d[(y,x)] = x

    for y in range(10):
        d = {}
        acc = 0
        ls = [str(x) for x in range(100000)]
        t0 = time.time()
        f(d, str(y), ls)
        acc += time.time() - t0
    print(f"Python\t{acc}")

python_bench()
numba_bench()
mixed_bench()
```

Results:

```
Python	0.019919395446777344
Numba	0.060488224029541016
Mixed	2.934454917907715
```

### CODE REVIEW

File: tests/regression/test_network_of_populations.py [hunk 1]
```diff
@@ -21,6 +21,8 @@ def _assert_baseline(result, force_update=False):
 
 def _assert_baseline_dataframe(result, force_update=False):
     with tempfile.NamedTemporaryFile(mode="w") as fp:
```
Comment (bobturneruk) [inline tests/regression/test_network_of_populations.py:23]: I know this is not changed by this PR, but could you add delete=False please for Windows compatibility?

File: tests/unit/test_network_of_populations/test_init.py [hunk 2]
```diff
@@ -412,7 +412,7 @@ def test_doInternalInfectionProcess_between_ages():
 
     new_infected = np.getInternalInfectiousContactsInNode(current_state, age_matrix, 1.0, ["I", "A"], True,
                                                           numpy.random.default_rng(123))
-    assert new_infected == {"m": 11, "o": 11}
+    assert new_infected == {"m": 12, "o": 13}
```
Comment (magicicada) [inline tests/unit/test_network_of_populations/test_init.py:415]: I don't quite understand why this change?  I'm sure I'm just not quite understanding something.

**Comment by bobturneruk** [comment]
What are the axes in the "Correctness evidence" plots? They're clearly very similar before and after, just wonder what they're showing...

File: simple_network_sim/network_of_populations/__init__.py [hunk 3]
```diff
@@ -536,61 +536,69 @@ def getInternalInfectiousContactsInNode(
...
     # Number of contacts cannot be higher than the number of people in the node
-    numberOfContacts = np.minimum(numberOfContacts, int(totalInAge))
-    numberOfContacts = random_state.hypergeometric(
-        int(susceptibles),
-        int(totalInAge) - int(susceptibles),
-        numberOfContacts
-    )
-    return cast(float, np.sum(numberOfContacts))
+    numbersOfContacts = np.minimum(np.array(numbersOfContacts).T, int(totalInAge))
+    result = random_state.hypergeometric(
```
Comment (FiodorG) [inline simple_network_sim/network_of_populations/__init__.py:596]: Does the call to hypergeometric work when numberOfContacts is 0? it didn't for me which is why I needed the filtering > 0 line but now I don't see it anymore?

File: simple_network_sim/network_of_populations/__init__.py [hunk 4]
```diff
@@ -536,61 +536,69 @@ def getInternalInfectiousContactsInNode(
...
 
-    numberOfContacts = random_state.poisson(contacts, size=int(infectious))
-    # If 0, the hypergeometric calls fails, but mathematically is 0
-    numberOfContacts = numberOfContacts[numberOfContacts > 0]
+    inf_max = int(round(max(infectious)))
+    # each np.array represent one original (contact, infectious) pair
+    numbersOfContacts: List[np.array] = []
+    for contact, infected in zip(contacts, infectious):
+        x = random_state.poisson(contact, int(round(infected)))
+        numbersOfContacts.append(np.concatenate([x, np.zeros(inf_max - len(x), dtype=int)]))
```
Comment (FiodorG) [inline simple_network_sim/network_of_populations/__init__.py:593]: What do you think about initializing a `np.zeros((inf_max, len(infection)))`, and then assigning into it row by row, rather than the more tedious list of arrays creation?

File: simple_network_sim/network_of_populations/__init__.py [hunk 5]
```diff
@@ -536,61 +536,69 @@ def getInternalInfectiousContactsInNode(
...
-    numberOfContacts = np.minimum(numberOfContacts, int(totalInAge))
-    numberOfContacts = random_state.hypergeometric(
-        int(susceptibles),
-        int(totalInAge) - int(susceptibles),
-        numberOfContacts
-    )
-    return cast(float, np.sum(numberOfContacts))
+    numbersOfContacts = np.minimum(np.array(numbersOfContacts).T, int(totalInAge))
+    result = random_state.hypergeometric(
+        [int(susceptibles)] * len(contacts),
```
Comment (FiodorG) [inline simple_network_sim/network_of_populations/__init__.py:597]: Same as above, why not using `np.full(len(contacts), int(susceptibles)))` rather than the list multiplication?

File: tests/unit/test_network_of_populations/test_init.py (hunk 2 above)
Comment (aflag) [reply tests/unit/test_network_of_populations/test_init.py:415]: The order of operations is now different, so the model results are different.

**Comment by aflag** [comment]
> What are the axes in the "Correctness evidence" plots? They're clearly very similar before and after, just wonder what they're showing...

Ah, I forgot to show the code I used to generate those. They histograms over `a` and `b`

```
    a = []
    b = []
    for seed in numpy.random.SeedSequence().generate_state(1000000):
        new_infected = np.getInternalInfectiousContactsInNode(current_state, age_matrix, 0.3, ["I", "A"], True, numpy.random.default_rng(seed))
        a.append(new_infected["m"])
        b.append(new_infected["o"])
```

They are essentially:
```
from matplotlib import pyplot as plt

plt.hist(a, bins=15)
plt.hist(b, bins=15)
```

So, the y-axis is a count and the x axis is the number of infectious contacts, as reported by getInternalInfectiousContactsInNode. I've repeated that 1,000,000 times for the same parameters and I got that histogram.

**Comment by aflag** [comment]
> I assume the changes in tests/regression/data/test_network_of_populations.py__test_basic_simulation_stochastic and tests/regression/data/test_sampleUseOfModel.py__test_stochastic_cli_run are very small? You've checked that it's the case?

I've checked it by plotting 

```
df[df.state.isin(["A", "I"])].groupby("date").sum().total.plot()
```

and I couldn't see the difference between plots.

File: simple_network_sim/network_of_populations/__init__.py (hunk 3 above)
Comment (aflag) [reply simple_network_sim/network_of_populations/__init__.py:596]: The hypergeometric implementation from `np.random.default_rng()` is ok with that:

```
>>> np.random.default_rng().hypergeometric(1,2, 0)
0
```

It's that small difference of implementation I told you about the other day.

File: simple_network_sim/network_of_populations/__init__.py (hunk 5 above)
Comment (aflag) [reply simple_network_sim/network_of_populations/__init__.py:597]: The reason is that I didn't know about `np.full` :P

Comment (aflag) [reply simple_network_sim/network_of_populations/__init__.py:597]:
```
timeit.timeit("[312] * 10", globals=globals())
0.10730258299736306
timeit.timeit("np.full(10, 312)", globals=globals())
1.6013900120160542
````
np.full seems slower, though.

File: simple_network_sim/network_of_populations/__init__.py (hunk 4 above)
Comment (aflag) [reply simple_network_sim/network_of_populations/__init__.py:593]: Sounds good.

File: tests/unit/test_network_of_populations/test_init.py (hunk 2 above)
Comment (FiodorG) [reply tests/unit/test_network_of_populations/test_init.py:415]: > I believe the numbers were just chosed to match whatever the model returned with the seed 123 previously
Just to confirm: indeed it was

File: simple_network_sim/network_of_populations/__init__.py (hunk 5 above)
Comment (FiodorG) [reply simple_network_sim/network_of_populations/__init__.py:597]: That's very surprising to me! However, one thing though, is that hypergeom might convert the lists to np arrays internally, in which case:
image

Because I am seeing this:
image

Comment (FiodorG) [reply simple_network_sim/network_of_populations/__init__.py:597]: Ref: https://stackoverflow.com/questions/5891410/numpy-array-initialization-fill-with-identical-values/44921249#44921249

Comment (FiodorG) [reply simple_network_sim/network_of_populations/__init__.py:597]: And for some reason that I don't fully understand, `np.zeros(100) + 312` is slightly faster too??

**Comment by aflag** [comment]
These are the result of running 

```
df = pd.read_csv("tests/regression/data/test_sampleUseOfModel.py__test_stochastic_cli_run")
df.groupby(["date","state"]).sum().reset_index().pivot(index="date", columns="state", values="total").plot()
```

on master vs on this branch. The graphs are visually identical.

Master:
before-change

issue-639 branch:
after-change
